packages/@aws-cdk/aws-codeartifact/lib/repository.ts

```diff
@@ -0,0 +1,404 @@
+
+  /**
+     * Grants the given IAM identity permissions to read/write from the repository
+     */
+  grantReadWrite(identity: iam.IGrantable): iam.Grant;
+
+  /**
+     * Defines a CloudWatch event rule which triggers for repository events. Use
+     * `rule.addEventPattern(pattern)` to specify a filter.
+     */
+  onEvent(id: string, options?: events.OnEventOptions): events.Rule;
+
+  /**
+     * Defines a CloudWatch event rule which triggers when a "CodeArtifact Package
+     *  Version State Change" event occurs.
+     */
+  onPackageVersionStateChange(id: string, options?: events.OnEventOptions): events.Rule;
+}
+
+/**
+ * Properties for a new CodeArtifact repository
+ */
+export interface RepositoryProps {
+  /**
+     * Name the repository
+     */
+  readonly repositoryName: string,
+  /**
+     * A text description of the repository.
+     * @default ''
+     */
+  readonly description?: string,
+  /**
+     * The name of the domain that contains the repository.
+     */
+  readonly domain: IDomain,
+  /**
+     * Upstream repositories for the repository
+     * @see https://docs.aws.amazon.com/codeartifact/latest/ug/repos-upstream.html
+     * @default None
+     */
+  readonly upstreams?: IRepository[],
+  /**
+     * An array of external connections associated with the repository.
+     * @default None
+     * @see https://docs.aws.amazon.com/codeartifact/latest/ug/external-connection.html#adding-an-external-connection
+     */
+  readonly externalConnections?: e.ExternalConnection[],
+
+  /**
+     * Principal to associate allow access to the repository
+     * @default AccountRootPrincipal
+     */
+  readonly principal?: iam.IPrincipal
+
+  /**
+   * Principal to associate allow access to the repository
+   * @default Read/Write
+   */
+  readonly policyDocument?: iam.PolicyDocument
+}
+
+/**
+ * Properties for a new CodeArtifact repository policy restricting package manipulation
+ */
+export interface PolicyRepositoryPackage {
+  /**
+     * Package format
+     */
+  readonly packageFormat: string,
+  /**
+     * Package namespace
+     */
+  readonly packageNamespace: string,
+  /**
+     * Package name
+     */
+  readonly packageName: string,
+}
+
+/**
+ * Either a new or imported Repository
+ */
+export abstract class RepositoryBase extends Resource implements IRepository {
+
+  /** @attribute */
+  public abstract readonly repositoryArn: string = '';
+  /** @attribute */
+  public abstract readonly repositoryName: string = '';
+  /** @attribute */
+  public abstract readonly repositoryDescription: string = '';
+  /** @attribute */
+  public abstract readonly repositoryDomainOwner: string = '';
+  /** @attribute */
+  public abstract readonly repositoryDomainName: string = '';
+
+  constructor(scope: Construct, id: string) {
+    super(scope, id, {});
+  }
+
+  /** Implement the {@link IRepository.grantRead} method. */
+  public grantRead(identity: iam.IGrantable): iam.Grant {
+    return iam.Grant.addToPrincipal({
```

Comment:
I believe this should probably be using `addToPrincipalOrResource`. This will allow adding to the Resource policy when the identity can't be added to (e.g., imported roles).

Also, consider a private `grant` method that contains the logic, and all of the `grant*` methods calling out to it with their actions.